### PR TITLE
Fix parsing of invalid timezone names

### PR DESCRIPTION
**Pull request by karadyauran**
Added some parsing requirements to avoid illegal zone ids

## Description
I have added the new parsing requirements in method `normalizeZoneId`. Now it checks the right format of time zone according to IANA's timezone database. 

## Motivation and Context
[Issue Link](https://github.com/prestodb/presto/issues/21461)

# Test

`given data -> received data`

```
1. zoneId = "etc/gmt+8" -> "+08:00"
2. zoneId = "etc/gmt-08:20" -> "-08:20"
3. zoneId = "etc/utc+1" -> "-01:00"
4. zoneId = "etc/utc-1" -> "+01:00"
5. zoneId = "etc/greenwich+01:00" -> "+01:00"
6. zoneId = "etc/greenwich+6" -> "+06:00"
7. zoneId = "etc/greenwich-6" -> "-06:00"
8. zoneId = "etc/greenwich+11:23" -> "+11:23"
9. zoneId = "etc/universal+08:00" -> "+08:00"
10. zoneId = "etc/universal-08:00" -> "-08:00"
11. zoneId = "etc/universal+11:23" -> "+11:23"
12. zoneId = "etc/uct+06:00" -> "+06:00"
13. zoneId = "etc/uct-10:10" -> "-10:10"
14. zoneId = "etc/uct+1" -> "+01:00"
15. zoneId = "etc/zulu+06:00" -> "+06:00"
16. zoneId = "etc/zulu-10:10" -> "-10:10"
17. zoneId = "etc/zulu+1" -> "+01:00"]
18. zoneId = "Zulu" -> "+00:00"
19. zoneId = "uct" -> "+00:00"
20. zoneId = "universal" -> "+00:00"
21. zoneId = "GMT-13:00" -> TimeZoneNotSupportedException
22. zoneId = "etc/6" -> TimeZoneNotSupportedException
23. zoneId = "etc/*" -> TimeZoneNotSupportedException
24. zoneId = "etc/" -> TimeZoneNotSupportedException
25. zoneId = "etc/gmt+24:00" -> TimeZoneNotSupportedException
26. zoneId = "etc/gmt+01" -> TimeZoneNotSupportedException
27. zoneId = "etc/gmt+01:000" -> TimeZoneNotSupportedException
28. zoneId = "etc/utc+27" -> TimeZoneNotSupportedException
29. zoneId = "etc/utc-13" -> TimeZoneNotSupportedException
30. zoneId = "etc/gmt-16" -> TimeZoneNotSupportedException
31. zoneId = "etc/greenwich+01" -> TimeZoneNotSupportedException
32. zoneId = "etc/universal+01" -> TimeZoneNotSupportedException
33. zoneId = "fsdkjflksdflsdlfj" -> TimeZoneNotSupportedException
34. zoneId = "ETC/+06:00" -> TimeZoneNotSupportedException
35. zoneId = "ETC/06:00" -> TimeZoneNotSupportedException
36. zoneId = "ETC/+6" -> TimeZoneNotSupportedException
```

## Release Notes
```
== RELEASE NOTES ==

General Changes
* Add new parameters `startsWithEtcGmt`, `startsWithEtc`, `startsWithEtcUtc`
* Deleted some logic components
```

@tdcmeehan

### CODE REVIEW

**Comment by karadyauran** [comment]
@tdcmeehan please, check my submission, when you have time 🙏

**Comment by tdcmeehan** [comment]
There is a checkstyle violation (see Maven Checks build output: ` src/main/java/com/facebook/presto/common/type/TimeZoneKey.java:[269] (regexp) RegexpMultiline: Multiple consecutive blank lines`)

Please also update the unit tests to test restrive behavior

**Comment by karadyauran** [comment]
@tdcmeehan if the format is wrong, should the program throw an exception?

**Comment by tdcmeehan** [comment]
> @tdcmeehan if the format is wrong, should the program throw an exception?

Yes

**Comment by karadyauran** [comment]
> There is a checkstyle violation (see Maven Checks build output: ` src/main/java/com/facebook/presto/common/type/TimeZoneKey.java:[269] (regexp) RegexpMultiline: Multiple consecutive blank lines`)
> 
> Please also update the unit tests to test restrive behavior

I have done all this tasks. Please, check it 🙏

**Comment by tdcmeehan** [comment]
I think we should add more tests that show the behavior is more restrictive.  #21461 had some examples that would make great test cases.

**Comment by karadyauran** [comment]
> I think we should add more tests that show the behavior is more restrictive.  #21461 had some examples that would make great test cases.

I have rebuilt the validation logic to more strictly version. But I still have bugs related to exception. Please, check this out 🙏

**Comment by tdcmeehan** [comment]
Thank you, but I still do not see any extra test cases.  Please add full test coverage and let me know when ready--thanks!

**Comment by karadyauran** [comment]
> Thank you, but I still do not see any extra test cases. Please add full test coverage and let me know when ready--thanks!

Done!

File: presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java [hunk 1]
```diff
@@ -1,140 +1,123 @@
...
-import javax.inject.Named;
-
-import java.lang.reflect.InvocationTargetException;
-import java.lang.reflect.Method;
-
-import static org.testng.Assert.*;
-
-public class TimeZoneKeyTest {
-
-  TimeZoneKey timeZoneKey = new TimeZoneKey("1", (short) 1);
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java]: Why is this change made?  Similarly for the visibility change in TimeZoneKey

File: presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java [hunk 2]
```diff
@@ -1,140 +1,123 @@
...
+
+ import javax.inject.Named;
+
+ import static org.testng.Assert.*;
+
+ public class TimeZoneKeyTest {
+
+   @Test
+   @Named("GMT TEST")
+   public void gmtZoneIdTest() {
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java:26]: Please look at CheckStyle output (maven-checks), and see README.md for formatting suggestings.  It looks like your formatting is off.

File: presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java [hunk 3]
```diff
@@ -1,140 +1,123 @@
...
+
+     zoneId = "etc/zulu+1";
+     normalizedZoneId = TimeZoneKey.normalizeZoneId(zoneId);
+     assertEquals(normalizedZoneId, "+01:00");
+   }
+
+   @Test
+   @Named("EXCEPTION TEST")
+   public void exceptionZoneIdTest() {
+     assertThrows(TimeZoneNotSupportedException.class, () -> TimeZoneKey.normalizeZoneId("GMT-13:00"));
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java:119]: Please keep the previous version if possible, I find it easier to read.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

File: presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java (hunk 3 above)
Comment (karadyauran) [reply presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java:119]: I need to return the second version for test cases and keep normalizeZoneId() method private ?

**Comment by karadyauran** [comment]
@tdcmeehan I have fixed all code style issues, but I still have an issue with the maven dependency problems. Please, check this out 🙏

File: presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+import javax.inject.Named;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import static org.assertj.core.api.Fail.fail;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class TimeZoneKeyTest
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TimeZoneKeyTest.java:28]: Can you please revert this?  Our tests start with Test, not end with it.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

File: presto-common/src/main/java/com/facebook/presto/common/type/TimeZoneKey.java [hunk 5]
```diff
@@ -210,107 +213,147 @@ public static boolean isUtcZoneId(String zoneId)
         return normalizeZoneId(zoneId).equals("utc");
     }
 
+    private static final Set<String> AVAILABLE_ZONE_IDS = ZoneId.getAvailableZoneIds().stream()
+            .map(String::toLowerCase)
+            .collect(Collectors.toSet());
+    private static final Pattern LONG_ETC_PATTERN = Pattern.compile("etc/(gmt|greenwich|uct|universal|utc|zulu)[+-]\\d{2}:\\d{2}");
```
Comment (tdcmeehan) [inline presto-common/src/main/java/com/facebook/presto/common/type/TimeZoneKey.java:219]: Put constants at the top

Comment (karadyauran) [reply presto-common/src/main/java/com/facebook/presto/common/type/TimeZoneKey.java]: Done!

**Comment by tdcmeehan** [comment]
Can you revert most of your changes in the test class?  The way I see it, many of the tests that previously succeeded should now fail, and we simply need to modify the assertions to show that.

**Comment by tdcmeehan** [comment]
Here is a patch of how I think the changes ought to go.  The changes in TimeZoneKey are difficult to read--https://gist.github.com/tdcmeehan/ae6927aa63ea3d81c70414fe8cb727e9

**Comment by karadyauran** [comment]
Do I understand correctly that the changes in the git are inconvenient to read because too many commits have been made? If so, I can recreate the branch and make the changes there in a new way.

If I look at the whole code and the tests that pass, is the logic of the written method correct?

**Comment by tdcmeehan** [comment]
Let's start from the test cases.  Please revert your changes to the test case, and make more individual changes.  For example, whereas this before would pass:

`assertSame(TimeZoneKey.getTimeZoneKey("etc/gmt+00:00"), UTC_KEY);`

Now it should throw:

`assertThrows(TimeZoneKey.getTimeZoneKey("etc/gmt+00:00"), ...`

Please just make small, individual changes to the tests, and we can work our backward from that.

**Comment by karadyauran** [comment]
I have understood your task and made all the changes! please, check these changes, and if you have any suggestions or adjustments, tell me and I will fix them.

File: presto-common/src/main/java/com/facebook/presto/common/type/TimeZoneKey.java [hunk 6]
```diff
@@ -212,105 +221,157 @@ public static boolean isUtcZoneId(String zoneId)
...
+        if (originalZoneId == null || originalZoneId.isEmpty()) {
+            throw new TimeZoneNotSupportedException("Zone ID cannot be null or empty");
+        }
+
         String zoneId = originalZoneId.toLowerCase(ENGLISH);
 
-        boolean startsWithEtc = zoneId.startsWith("etc/");
-        if (startsWithEtc) {
-            zoneId = zoneId.substring(4);
+        if (isFixedOffsetTimeZone(zoneId) || isUtcEquivalentName(zoneId)) {
```
Comment (tdcmeehan) [inline presto-common/src/main/java/com/facebook/presto/common/type/TimeZoneKey.java:230]: This seems to be throwing for UTC and GMT.  Doesn't seem expected?

File: presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java [hunk 7]
```diff
@@ -46,89 +46,87 @@ public void testUTC()
...
-        assertSame(TimeZoneKey.getTimeZoneKey("etc/utc"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("etc/gmt"), UTC_KEY);
+        assertTimeZoneNotSupported("Z");
+        assertTimeZoneNotSupported("Zulu");
+        assertTimeZoneNotSupported("zulu");
+        assertTimeZoneNotSupported("ZULU");
+        assertTimeZoneNotSupported("UT");
+        assertTimeZoneNotSupported("UCT");
+        assertTimeZoneNotSupported("Universal");
+        assertTimeZoneNotSupported("GMT");
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java:56]: This should be supported

File: presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java [hunk 8]
```diff
@@ -46,89 +46,87 @@ public void testUTC()
...
     @Test
     public void testHourOffsetZone()
     {
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT0"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT+0"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT-0"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT+0"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT-0"), UTC_KEY);
+        assertTimeZoneNotSupported("GMT0");
+        assertTimeZoneNotSupported("GMT+0");
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java:79]: This should be supporte.  https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

**Comment by karadyauran** [comment]
I have this error in test `TimeZoneNotSupported Time zone not supported: GMT-13:00`.

In my opinion, it's a wrong time zone format, that is not supported by IANA's timezone database. The right format is `Etc/GMT+13`.

File: presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java [hunk 9]
```diff
@@ -47,88 +47,79 @@ public void testUTC()
...
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT+00:00"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT-00:00"), UTC_KEY);
+        assertTimeZoneNotSupported("GMT+00:00");
+        assertTimeZoneNotSupported("GMT-00:00");
         assertSame(TimeZoneKey.getTimeZoneKey("+00:00"), UTC_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("-00:00"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("etc/utc"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("etc/gmt"), UTC_KEY);
+        assertTimeZoneNotSupported("etc/utc");
+        assertTimeZoneNotSupported("etc/gmt");
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java:59]: Both of these entries are found here and should be supported: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

File: presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java [hunk 10]
```diff
@@ -47,88 +47,79 @@ public void testUTC()
...
         assertSame(TimeZoneKey.getTimeZoneKey("etc/gmt+00:00"), UTC_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("etc/gmt-00:00"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("etc/ut"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("etc/UT"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("etc/UCT"), UTC_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("etc/Universal"), UTC_KEY);
+        assertTimeZoneNotSupported("etc/ut");
+        assertTimeZoneNotSupported("etc/UT");
+        assertTimeZoneNotSupported("etc/UCT");
+        assertTimeZoneNotSupported("etc/Universal");
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java:66]: Both of these entries are found here and should be supported: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

File: presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java [hunk 11]
```diff
@@ -47,88 +47,79 @@ public void testUTC()
...
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT-0"), UTC_KEY);
         assertTimeZoneNotSupported("GMT7");
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT+7"), PLUS_7_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT-7"), MINUS_7_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT+7"), PLUS_7_KEY);
-        assertSame(TimeZoneKey.getTimeZoneKey("GMT-7"), MINUS_7_KEY);
+        assertTimeZoneNotSupported("GMT+7");
+        assertTimeZoneNotSupported("GMT-7");
+        assertTimeZoneNotSupported("GMT+7");
+        assertTimeZoneNotSupported("GMT-7");
```
Comment (tdcmeehan) [inline presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java:81]: These should be supported
Screenshot 2024-01-09 at 10 47 21 AM

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

File: presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java (hunk 11 above)
Comment (karadyauran) [reply presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java:81]: I think "GMT+7" is not the right time zone format. The right is "Etc/GMT+7".

Comment (tdcmeehan) [reply presto-common/src/test/java/com/facebook/presto/common/type/TestTimeZoneKey.java:81]: Yep, you're right!

**Comment by tdcmeehan** [comment]
Tests look good now, next will review function changes.

**Comment by karadyauran** [comment]
@tdcmeehan have you already reviewed my code?

**Comment by karadyauran** [comment]
Hello, Timothy!
Let me know, have you already looked at my code? And do I need to make a pull request on the other branch? 
Best, 
Alex

**Comment by tdcmeehan** [comment]
I have a large backlog of reviews.  Expect a response next week--appreciate your patience!  Feel free to ping on #dev if you want a review sooner.

**Comment by karadyauran** [comment]
Hi! Have you already reviewed my code?

**Comment by karadyauran** [comment]
@tdcmeehan Hey! What about my code?

**Comment by tdcmeehan** [comment]
@karadyauran looks better.  But I think we need to make the code switched by a flag to account for people who are relying on the old behavior.  Then, after a period of time, we can deprecate the old behavior.  Can you make it so that the new requirements are switched based off of a config flag?

**Pull request closed without merging**